source/language/openpulse.rst

```diff
@@ -0,0 +1,650 @@
+    rxchannel cap0 = rxch($0, "capture");
+    rxchannel cap1 = rxch($1, "capture");
+
+Frames
+------
+
+When interacting with qubits, it is powerful to track of a frame of reference, akin to the rotating
+frame of a Hamiltonian, throughout the execution of a program. Openpulse provides the ``frame``
+type which is responsible for tracking two properties:
+
+- Tracking time appropriately so programs do not need to deal in absolute time or with the
+  bookkeeping of advancing time in a sequence of pulses.
+- Tracking phase by producing a complex value given an input time (i.e. via the mathematical
+  relationship :math:`e^{i\left(2\pi f t + \theta\right)}`,  where `f` is frequency and
+  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be
+  defined in the rotating frame with the effect being an equivalent application in the lab
+  frame (i.e. with the carrier supplied by the ``frame``). Another motivation is to more naturally
+  implement a "virtual Z-gate", which does not require a physical pulse but rather shifts the phase
+  of all future pulses on that frame.
+
+The frame is composed of three parts:
+
+1. A frequency ``frequency`` of type ``float``.
+2. A phase ``phase`` of type ``angle``.
+3. A time which is manipulated implicitly and cannot be modified other
+   than through existing timing instructions like ``delay`` and ``barrier``. The time increment
+   is determined by the channel on which the frame is played (see :ref:`Timing` section).
+
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single NCO).
+
+Frame Construction
+~~~~~~~~~~~~~~~~~~
+
+Frames can be constructed using the ``Frame`` command e.g.
+
+.. code-block:: javascript
+
+  frame driveframe = Frame(5e9, 0.0); // Frame(float[size] frequency, angle[size] phase)
+
+would construct a frame with frequency `f`  5 GHz and phase :math:`\theta` 0.0. When
+instantiated, the frame time starts at 0. ``Frame``s can also be copied using the
+``copy`` command
+
+.. code-block:: javascript
+
+  frame driveframe1 = Frame(5e9, 0.0);
+  frame driveframe2 = copy(driveframe1);
+  driveframe2.phase = driveframe1.phase + pi/2;
+
+This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` with the same
+time as `driveframe1`.
+
+Unless a ``calconfig`` block is used to instantiate frames in global scope, a ``frame`` is
+locally scoped . To allow for parameterized access to globally defined frames, we can use
+the ``setframe`` and ``getframe`` functions
+
+.. code-block:: javascript
+
+   setframe(str name, frame fr) $q;
+   getframe(str name) $q-> frame;
+
+For example, we can parmaeterize frames on different qubits using the wildcard reference
+.. code-block:: javascript
+
+   calconfig {
```

Comment:
This offers a solution on scoping. This is a point for discussion...

Reply:
Review `getframe/setframe` vs `getattr/setattr`, consider qubit wildcards, etc...

Reply:
On a related note, are we also considering the ability to allow for other calibration parameters to skip across `defcal`s?  For example, should I be able to access a calibrated Rabi pulse or qubit frequency?

Reply:
I think this is a good idea. 

More generally, if we want feature parity with the current OpenPulse, we will need to allow other classical parameters as input to `defcal`'s. An example is doing a Rabi or spectroscopy experiment (input amp or freq). I am inclined to think of these as pulse subroutines rather than gate calibrations. They could then be called directly without reference to a gate. Maybe we should allow this with a different keyword?

Reply:
After discussion w/ the group, @stevenheidel and @taalexander are going to word towards adding a scoping construct analogous to `box` which may contain inline pulse statements. This allows us to distinguish between two use cases: calibrating a unitary gate and working at a low level of pulse control, at which time the IR is closer to the hardware.

Reply:
This has been added as the `cal` block implementation. Closing this thread. Thx @stevenheidel, @taalexander!